[BugFix] Be crash when restore primary key table cause by col unique id (#23383)

srlch opened the pull request:
## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes:
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #23383

## Problem Summary(Required):
Problem:
For a primary key table, if the table's col unique id is not not consecutive, be will crash when restore this table cause by the difference between segment file and tablet schema.

Solution:
reset tablet schema using schema in snapshot data

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

trueeyu commented on be/src/storage/tablet_manager.cpp:
```diff
@@ -1510,6 +1510,102 @@ Status TabletManager::create_tablet_from_meta_snapshot(DataDir* store, TTabletId
+    if (!tablet->init().ok()) {
+        LOG(WARNING) << "Fail to init restored tablet " << tablet_id;
+        wb.Clear();
+        RETURN_IF_ERROR(TabletMetaManager::clear_del_vector(store, &wb, tablet_id));
```
If clear_del_vector failed, will return error without clear other items such as rowset, delta_column_group ?

trueeyu commented on be/src/storage/tablet_manager.cpp:
```diff
@@ -1510,6 +1510,102 @@ Status TabletManager::create_tablet_from_meta_snapshot(DataDir* store, TTabletId
+    std::set<uint32_t> set2;
+    auto& v = snapshot_meta->tablet_meta().updates().versions(0).rowsets();
+    set2.insert(v.begin(), v.end());
+    if (!(set1 == set2)) {
```
set1 != set2?

trueeyu commented on be/src/storage/tablet_manager.cpp:
```diff
@@ -1510,6 +1510,102 @@ Status TabletManager::create_tablet_from_meta_snapshot(DataDir* store, TTabletId
+    auto tablet_meta = std::make_shared<TabletMeta>();
+    tablet_meta->init_from_pb(&snapshot_meta->tablet_meta());
+    if (tablet_meta->tablet_state() == TABLET_SHUTDOWN) {
+        LOG(INFO) << "Fail to load snapshot from " << schema_hash_path << ": tablet has been shutdown";
```
LOG(WARN) or Log(INFO)

trueeyu commented on be/src/storage/tablet_manager.cpp:
```diff
@@ -1510,6 +1510,102 @@ Status TabletManager::create_tablet_from_meta_snapshot(DataDir* store, TTabletId
+        return Status::NotFound("tablet path not exists");
+    }
+
+    std::unique_lock l(_get_tablets_shard_lock(tablet_id));
```
Will it lock for long time?

sevev commented on be/src/storage/tablet_manager.cpp:
```diff
@@ -1510,6 +1510,102 @@ Status TabletManager::create_tablet_from_meta_snapshot(DataDir* store, TTabletId
+        LOG_IF(WARNING, !st.ok()) << "Fail to clear meta store: " << st;
+        return Status::InternalError("tablet init failed");
+    }
+    auto st = _add_tablet_unlocked(tablet, true, true);
```
Is it enough to hold the tablet_shard lock here?

srlch committed:
[BugFix] Be crash when restore primary key table cause by col unique id (#23383)

Problem:
For a primary key table, if the table's col unique id is not not consecutive,
be will crash when restore this table cause by the difference between segment file and tablet schema.

Solution:
reset table schema in primary_key_move

Signed-off-by: srlch <linzichao@starrocks.com>

wanpengfei-git commented:
@Mergifyio backport branch-3.0

wanpengfei-git commented:
@Mergifyio backport branch-2.5
